include/mango_v3.hpp

```diff
@@ -197,6 +203,247 @@ struct EventQueue {
+  LeafNode,
+  FreeNode,
+  LastFreeNode
+};
+
+struct AnyNode {
+  NodeType tag;
+  uint8_t padding[BOOK_NODE_SIZE - 4];
+};
+
+struct InnerNode {
+  NodeType tag;
+  uint32_t prefixLen;
+  __uint128_t key;
+  uint32_t children[2];
+  uint8_t padding[BOOK_NODE_SIZE - 32];
+};
+
+struct LeafNode {
+  NodeType tag;
+  uint8_t ownerSlot;
+  uint8_t orderType;
+  uint8_t version;
+  uint8_t timeInForce;
+  __uint128_t key;
+  solana::PublicKey owner;
+  uint64_t quantity;
+  uint64_t clientOrderId;
+  uint64_t bestInitial;
+  uint64_t timestamp;
+};
+
+struct FreeNode {
+  NodeType tag;
+  uint32_t next;
+  uint8_t padding[BOOK_NODE_SIZE - 8];
+};
+
+struct Order {
+  Order(uint64_t price, uint64_t quantity) : price(price), quantity(quantity) {}
+  uint64_t price = 0;
+  uint64_t quantity = 0;
+};
+
+struct L1Orderbook {
+  uint64_t highestBid = 0;
+  uint64_t highestBidSize = 0;
+  uint64_t lowestAsk = 0;
+  uint64_t lowestAskSize = 0;
+  double midPoint = 0.0;
+  double spreadBps = 0.0;
+
+  bool valid() const {
+    return ((highestBid && lowestAsk) && (lowestAsk > highestBid)) ? true
+                                                                   : false;
+  }
+};
+
+class BookSide {
+ public:
+  BookSide(Side side) : side(side) {}
+
+  bool handleMsg(const nlohmann::json &msg) {
+    // ignore subscription confirmation
+    const auto itResult = msg.find("result");
+    if (itResult != msg.end()) {
+      return false;
+    }
+
+    const std::string encoded = msg["params"]["result"]["value"]["data"][0];
+    const std::string decoded = solana::b64decode(encoded);
+    return update(decoded);
+  }
+
+  Order getBestOrder() const {
+    return (!orders->empty()) ? orders->front() : Order(0, 0);
+  }
+
+  uint64_t getVolume(uint64_t price) const {
+    if (side == Side::Buy) {
+      return getVolume<std::greater_equal<uint64_t>>(price);
+    } else {
+      return getVolume<std::less_equal<uint64_t>>(price);
+    }
+  }
+
+ private:
+  template <typename Op>
+  uint64_t getVolume(uint64_t price) const {
+    Op operation;
+    uint64_t volume = 0;
+    for (auto &&order : *orders) {
+      if (operation(order.price, price)) {
+        volume += order.quantity;
+      } else {
+        break;
+      }
+    }
+    return volume;
+  }
+
+  bool update(const std::string decoded) {
+    if (decoded.size() != sizeof(BookSideRaw)) {
+      throw std::runtime_error("invalid response length " +
+                               std::to_string(decoded.size()) + " expected " +
+                               std::to_string(sizeof(BookSideRaw)));
+    }
+    memcpy(&(*raw), decoded.data(), sizeof(BookSideRaw));
+
+    auto iter = BookSide::BookSideRaw::iterator(side, *raw);
+    std::vector<Order> newOrders;
+    while (iter.stack.size() > 0) {
+      if ((*iter).tag == NodeType::LeafNode) {
+        const auto leafNode =
+            reinterpret_cast<const struct LeafNode *>(&(*iter));
+        const auto now = std::chrono::system_clock::now();
```

Comment:
you dont want to recalculate the current time inside the loop

Reply:
https://github.com/mschneider/solcpp/blob/9deb4c90abec347db7cdf288c7cc2d603d65b7aa/examples/ticker.cpp#L36
mango gravy from the initial example lol. 
pls create a seperate bug/issue for this.